### ISSUE TITLE
Client-side update card upon input changes or button presses (statically pre-loaded, cascading dropdowns, etc)

### ISSUE THREAD

**Issue by productboard-ac** [microsoft/AdaptiveCards#3112]
# Summary

Card authors have expressed the need for their card to change client-side (no server calls) as a result of user selecting/changing inputs, user pressing a button, etc.

# Scenarios

## Add new project

Some examples would be a dropdown that has an "Add new" option, when you select that, want to show a text box for adding a new item

![AddNew](https://user-images.githubusercontent.com/13246069/60295126-d4609200-98d7-11e9-9781-4238124c5bef.gif)

## Survey dissatisfaction

Or in surveys, when they select "Dissatisfied", you might want to have an input appear that asks them why they're dissatisfied, all client-side

![SurveyDissatisfied](https://user-images.githubusercontent.com/13246069/60295224-070a8a80-98d8-11e9-9008-0ae8803d19a4.gif)

## Changes based on checkbox

Another example someone mentioned was if the user checks a checkbox, they want to make some other fields required, so changing those other inputs to be isRequired.

## Add new row button

Another (fictitious) example is having an "Add new row" button that client-side adds a new input row when clicked

![StaticDynamic 1](https://user-images.githubusercontent.com/13246069/60295346-594bab80-98d8-11e9-8d1d-9ba24a2d811e.gif)

# Requirements

- P0 - When certain ChoiceSet value selected, show some elements, otherwise hide them

- P1 - When checkbox is selected, make certain inputs required

- P2 - Supporting a button that can add additional input rows (P2 since don't have any actual asks for this yet)

# Spec/schema

After the proposal process, update this with a link to the spec PR.

# Proposal comparisons

Category | Proposal 1 | Proposal 2 | Proposal 3
--|--|--|--
**Description** | Use live templating | Use onSelected+Action.Refresh | Use onSelected+Action.ToggleVisibility
**Ease of use** | Intuitive, easy | A bit cumbersome | A bit cumbersome
**Can ship in 1.3?** | No, dependent not only on templating but also live templating | Maybe, blocked on Action.Refresh | Yes
**Works for P0 ChoiceSet value selected show elements?** | Yes | Yes | Yes
**Works for P1 checkbox selected make inputs required?** | Yes | Yes | No... You could toggle visibility between a non-required input and a required input, but doing so would lose the input value they had previously typed since it'd be a different input with a different id
**Works for P2 add input row?** | Would need cardscript+data manipulations | No (while you can add items, you need to generate unique input ids which wouldn't be possible)... So would need cardscript+dom manipulations | Would need cardscript+dom manipulations

# Proposal 1: Use live templating

If we had live templating, this becomes pretty easy

## Examples

![AddNew](https://user-images.githubusercontent.com/13246069/60295126-d4609200-98d7-11e9-9781-4238124c5bef.gif)

```
{
  "type": "AdaptiveCard",
  "version": "2.0",
  "body": [
    {
      "type": "TextBlock",
      "text": "Project"
    },
    {
      "type": "Input.ChoiceSet",
      "id": "project",
      "style": "compact",
      "isMultiSelect": false,
      "choices": [
        {
          "$data": "{projects}",
          "title": "{$data}",
          "value": "{$data}"
        },
        {
          "title": "Add new project",
          "value": "__new__"
        }
      ]
    },
    {
      "$when": "{$inputs.project.value == '__new__'}",
      "type": "TextBlock",
      "text": "New project name"
    },
    {
      "$when": "{$inputs.project.value == '__new__'}",
      "type": "Input.Text",
      "id": "newProjectName"
    },
    {
      "type": "TextBlock",
      "text": "New task name"
    },
    {
      "type": "Input.Text",
      "id": "taskName"
    }
  ]
}
```

![SurveyDissatisfied](https://user-images.githubusercontent.com/13246069/60295224-070a8a80-98d8-11e9-9008-0ae8803d19a4.gif)

```
{
  "type": "AdaptiveCard",
  "version": "1.2",
  "body": [
    {
      "type": "TextBlock",
      "text": "How satisfied are you?"
    },
    {
      "type": "Input.ChoiceSet",
      "id": "satisfaction",
      "style": "compact",
      "isMultiSelect": false,
      "choices": [
        {
          "title": "Satisfied",
          "value": "satisfied"
        },
        {
          "title": "Neutral",
          "value": "neutral"
        },
        {
          "title": "Dissatisfied",
          "value": "dissatisfied"
        }
      ]
    },
    {
      "$when": "{$inputs.satisfaction.value == 'dissatisfied'}",
      "type": "TextBlock",
      "text": "We're sorry you're dissatisfied. Tell us why."
    },
    {
      "$when": "{$inputs.satisfaction.value == 'dissatisfied'}",
      "type": "Input.Text",
      "id": "explanation"
    }
  ]
}
```

![AddingPassengers](https://user-images.githubusercontent.com/13246069/60304899-a424ed80-98ef-11e9-8feb-0891f97ae71b.gif)

Not sure how this adding new row one would work. I think we would need a new `Action.ModifyData` action that would be able to add a new value to the data (we'd also need to be able to remove a value too, since people would want a "Delete row" button too).

Or if we supported author-provided JavaScript that can manipulate the data, that becomes even easier and probably a better experience

```
{
  "type": "AdaptiveCard",
  "version": "1.0",
  "body": [
    {
      "type": "TextBlock",
      "text": "Passengers"
    },
    {
      "$data": "{passengers}",
      "type": "Input.Text",
      "title": "New Input.Toggle",
      "id": "passenger{$index}",
      "placeholder": "Name"
    },
    {
      "type": "Action.RunScript",
      "title": "Add passenger",
      "script": "addPassenger()"
    }
  ]
}
```

**Author-provided script**

```
function addPassenger() {
    var passengers = getData().passengers;
    passengers.push({});

    setData({
        passengers: passengers
    });
}
```

## Host Config

Quickly describe changes to the host config, if any.

## Down-level impact

Describe how down-level is impacted for authors.

## Host burden

Describe the burden hosts have when we introduce this feature. For example, are they going to have to add something new to their host config or native styling as a result of a new element we're adding? Are there breaking API changes that hosts have to use? Are we deprecating something from host config that could cause potential confusion?

# Proposal 2: Use RefreshCard

Maybe we could use the RefreshCard concept to update the card?

# Proposal 3: Use Action.ToggleVisibility

Maybe we could use Action.ToggleVisibility?

![SurveyDissatisfied](https://user-images.githubusercontent.com/13246069/60295224-070a8a80-98d8-11e9-9008-0ae8803d19a4.gif)

```
{
  "type": "AdaptiveCard",
  "version": "1.2",
  "body": [
    {
      "type": "TextBlock",
      "text": "How satisfied are you?"
    },
    {
      "type": "Input.ChoiceSet",
      "id": "satisfaction",
      "style": "compact",
      "isMultiSelect": false,
      "choices": [
        {
          "title": "Satisfied",
          "value": "satisfied",
          "onSelected": {
            "type": "Action.ToggleVisibility",
            "elements": [
              {
                "elementId": "explanationContainer",
                "isVisible": false
              }
            ]
          }
        },
        {
          "title": "Neutral",
          "value": "neutral",
          "onSelected": {
            "type": "Action.ToggleVisibility",
            "elements": [
              {
                "elementId": "explanationContainer",
                "isVisible": false
              }
            ]
          }
        },
        {
          "title": "Dissatisfied",
          "value": "dissatisfied",
          "onSelected": {
            "type": "Action.ToggleVisibility",
            "elements": [
              {
                "elementId": "explanationContainer",
                "isVisible": true
              }
            ]
          }
        }
      ]
    },
    {
      "type": "Container",
      "id": "explanationContainer",
      "items": [
        {
          "type": "TextBlock",
          "text": "We're sorry you're dissatisfied. Tell us why."
        },
        {
          "type": "Input.Text",
          "id": "explanation"
        }
      ]
    }
  ]
}
```

The redundancy could be eliminated by using templating

```
{
  "type": "AdaptiveCard",
  "version": "1.2",
  "body": [
    {
      "type": "TextBlock",
      "text": "How satisfied are you?"
    },
    {
      "type": "Input.ChoiceSet",
      "id": "satisfaction",
      "style": "compact",
      "isMultiSelect": false,
      "choices": [
        {
          "$data": {
            "choices": [
              {
                "title": "Satisfied",
                "value": "satisfied"
              },
              {
                "title": "Neutral",
                "value": "neutral"
              },
              {
                "title": "Dissatisfied",
                "value": "dissatisfied"
              }
            ]
          },
          "title": "{title}",
          "value": "{value}",
          "onSelected": {
            "type": "Action.ToggleVisibility",
            "elements": [
              {
                "elementId": "explanationContainer",
                "isVisible": "{value == 'dissatisfied'}"
              }
            ]
          }
        }
      ]
    },
    {
      "type": "Container",
      "id": "explanationContainer",
      "items": [
        {
          "type": "TextBlock",
          "text": "We're sorry you're dissatisfied. Tell us why."
        },
        {
          "type": "Input.Text",
          "id": "explanation"
        }
      ]
    }
  ]
}
```

![AddingPassengers](https://user-images.githubusercontent.com/13246069/60304899-a424ed80-98ef-11e9-8feb-0891f97ae71b.gif)

Adding a new row would probably require scripting code similar to the first proposal

**Comment by productboard-ac** [comment]
Link to feature: https://adaptivecards.productboard.com/feature-board/planning/features/1772148

**Comment by shalinijoshi19** [comment]
@paulcam206 assigning to you for driving. thanks!
